### ISSUE TITLE
The list grows infinitely until the app crashes with OOM

### ISSUE THREAD

**Issue by Nek-12** [Ahmad-Hamwi/lazy-pagination-compose#4]
With current implementation, after scrolling through several thousand items, the app crashes with OOM because it fails to allocate a huge list of items. Before that, the app slows down to a halt due to GC pressure.

**Comment by Ahmad-Hamwi** [comment]
Thank you for your observation, may I ask you to test this on a normal list? Not a paginated one. I need first to know whether this is a compose problem or the library's problem.

Also let me know what platform you're on.

**Comment by Nek-12** [comment]
The same list if not paginated will also crash the app.

**Comment by Ahmad-Hamwi** [comment]
This means this isn't this library's bug, it is probably a misuse of the lazy list or an actual bug on compose's side.

Are you assigning keys to each item in the list? If not, do it and let me know if it helped. The key should be unique across all items.

**Comment by Nek-12** [comment]
This is a library bug on this line https://github.com/Ahmad-Hamwi/lazy-pagination-compose/blob/104dbe2fd84312f87e85e8e3b5c18c845bb4aead/lazy-pagination-compose/src/commonMain/kotlin/io/github/ahmad_hamwi/compose/pagination/PaginationState.kt#L49. You grow the list infinitely

**Comment by Ahmad-Hamwi** [comment]
True, otherwise your data will be lost, you won't be able to go back to your previously scrolled data, this isn't the behavior I expect when scrolling in any feed app for example :)

If you suggest any better way of holding the data without a loss, let me know.

**Comment by Ahmad-Hamwi** [comment]
You're talking about disk caching, now it makes sense.

I think supporting such optimization will take some time, especially on each platform, I may start with Android.

Will consider such change and I'm open to any contributions.

**Comment by Nek-12** [comment]
No I'm not talking about disk caching, I meant an in-memory pool.

**Comment by Ahmad-Hamwi** [comment]
I would also appreciate a minimal sample project that has produces the crash.

**Comment by Nek-12** [comment]
> Can you provide me with resources on such in-memory object pool? A quick search showed me that it can be the same as an in-memory DB, but I'm not sure how it differs from using heap directly (current implemention) in terms of optimization.

https://android.googlesource.com/platform/frameworks/support/+/androidx-paging-release/paging/paging-common/src/commonMain/kotlin/androidx/paging/ItemSnapshotList.kt

**Comment by terrakok** [comment]
@Nek-12 instead of the list of objects you can keep paged list of IDs. And every item will load its data by its own. (and will free the memory when the view leave the composition)
